File: Bai_51/main.py
while False: # Phứt tạp
    import os
    with open("E:\\Project Visual Studio Code\\Summary_exercises\\Bai_51\\CUM.INP", "r") as ifile:
        s = ifile.readline().strip()

    with open("E:\\Project Visual Studio Code\\Summary_exercises\\Bai_51\\CUM.TXT", "w") as otempfile:
        for i in range(len(s) - 1):
            cnt_start = 0
            cnt_end = 0
            index_start = 0
            index_end = 0

            if s[i] == '(':
                cnt_start += 1
                index_start = i
                for j in range(i + 1, len(s)):
                    if s[j] == '(':
                        cnt_start += 1
                    if s[j] == ')':
                        cnt_end += 1
                        if cnt_start == cnt_end:
                            index_end = j
                            break
                otempfile.write(s[index_start:index_end + 1] + '\n')

    with open("E:\\Project Visual Studio Code\\Summary_exercises\\Bai_51\\CUM.TXT", "r") as itempfile:
        word_length = {}
        for line in itempfile:
            word = line.strip()
            word_length[word] = len(word)

        sorted_word_length = sorted(word_length.items(), key=lambda x: x[1])
    os.remove("E:\\Project Visual Studio Code\\Summary_exercises\\Bai_51\\CUM.TXT")

    with open("E:\\Project Visual Studio Code\\Summary_exercises\\Bai_51\\CUM.OUT", "w") as ofile:
        cnt = s.count('(')
        ofile.write(str(cnt) + '\n')
        for word, length in sorted_word_length:
            ofile.write(word + '\n')

while True: # Tối ưu và đơn giản
    with open("CUM.INP", "r") as ifile:
        s = ifile.readline().strip()
        idx = []
        polynomials = []
        cnt = 0

        for i in range(len(s)):
            if s[i] == '(':
                idx.append(i)
                cnt += 1
            if s[i] == ')':
                cnt -= 1
                st = idx.pop()
                tmp = s[st:i+1]
                polynomials.append(tmp)
        with open("CUM.OUT", "w") as ofile:
            if cnt == 0:
                ofile.write(str(len(polynomials)) + '\n')
                for poly in polynomials:
                    ofile.write(str(poly) + '\n')
            else:
                ofile.write(str(-1) + '\n')

File: Bai_51/main.cpp
#include<bits/stdc++.h>
using namespace std;
int main()
{
    ifstream ifile("CUM.INP");
    string s;
    getline(ifile,s);
    ifile.close();

    int cnt = 0;
    ofstream otempfile("CUM.TXT");
    for(int i = 0; i < s.length() - 1; i++)
    {
        int cnt_start = 0;
        int cnt_end = 0;
        int index_start = 0;
        int index_end = 0;
        if(s[i] == '(')
        {   
            cnt ++;
            cnt_start++;
            index_start = i;
            for(int j = i + 1; i < s.length(); j++)
            {
                if(s[j] == '(')
                    cnt_start++;
                if(s[j] == ')')
                {
                    cnt_end++;
                    if(cnt_start == cnt_end)
                    {
                        index_end = j;
                        break;
                    } 
                }
            }
            for(int m = index_start; m <= index_end; m++)
                otempfile << s[m];
            otempfile << endl;
        } 
    }
    otempfile.close();

    #if 0 // Dùng map và vector
    ifstream itempfile("CUM.TXT");
    string word;
    map<string,int> word_length;
    while (itempfile >> word)
        word_length[word] = word.length();
    itempfile.close();
    remove("CUM.TXT");

    vector<pair<string, int>> sorted_word_length(word_length.begin(), word_length.end());
    sort(sorted_word_length.begin(), sorted_word_length.end(), [](const pair<string, int>& a, const pair<string, int>& b){
        return a.second < b.second; 
    });

    ofstream ofile("CUM.OUT");
    ofile << cnt << endl;
    for (auto it : sorted_word_length)
        ofile << it.first << endl;
    ofile.close();
    #endif

    #if 0 // Dùng map
    ifstream itempfile("CUM.TXT");
    string word;
    map<pair<int,string>,bool> word_length;
    while (itempfile >> word)
        word_length[make_pair(word.length(),word)] = true;
    itempfile.close();
    remove("CUM.TXT");

    ofstream ofile("CUM.OUT");
    ofile << cnt << endl;
    for (auto it : word_length)
        ofile << it.first.second << endl;
    ofile.close();
    #endif

    #if 0 // Dùng set
    ifstream itempfile("CUM.TXT");
    string word;
    set<pair<int,string>> word_length;
    while (itempfile >> word)
        word_length.insert(make_pair(word.length(),word));
    itempfile.close();
    remove("CUM.TXT");

    ofstream ofile("CUM.OUT");
    ofile << cnt << endl;
    for (auto it : word_length)
        ofile << it.second << endl;
    ofile.close();
    #endif
    
    return 0;
}
